### ISSUE TITLE
Transformer early stop scheme

### ISSUE THREAD

**Issue by cshanbo** [tensorflow/tensor2tensor#131]
Hi all,
I've recently done some tests on `transformer`, translation tasks, and the experimental results are exciting!

I also noticed that, when I set the `training_steps` to a large number, e.g, 300000, the training process cannot reach the specified steps, while the `loss` is decreasing. There was no exception raised. I didn't see any related information in the log. 

My questions are:
1. Is it `early stop`?
2. If yes, is it managed automatically by `tf.contrib.learn.Experiment` (Sorry for asking this because I'm not so familiar with this api, and didn't see any related settings in the codes)? 
3. How can I "fix" this? 

Following is a piece of log.

```
INFO:tensorflow:loss = 2.29189, step = 116922 (208.006 sec)
INFO:tensorflow:Saving checkpoints for 116952 into model.ckpt.
INFO:tensorflow:global_step/sec: 0.45337
skipped ......
INFO:tensorflow:loss = 2.52611, step = 118522 (208.300 sec)
INFO:tensorflow:global_step/sec: 0.480134
INFO:tensorflow:loss = 1.96562, step = 118622 (208.275 sec)
INFO:tensorflow:Saving checkpoints for 118651 into model.ckpt.
```

**Comment by lukaszkaiser** [comment]
I don't think we have any early stopping scheme, and I don't think there is anything like this in `Estimator` or `Experiment`. The flag is `--train_steps` though, not `--training_steps`, you did set the right one? If you did, it shouldn't be stopping, and if it does then it's a bug. Can you reproduce?

**Comment by cshanbo** [comment]
Yes. the flag is `--train_steps`. 
It's not 100% reproducible. It happened twice in my experiments.

**Comment by lukaszkaiser** [comment]
Can you still see it? I cannot reproduce this at all for now, let us know if you have a way of making it happen at least quite probably.

**Comment by cshanbo** [comment]
I saw it again. But due to the fact that no related information in log, I'm not sure why it corrupted.
The first time I saw this issue, my settings were:

Model   |  vocab  | steps      |  cards
----------|------------|-------------|---------------
transformer_big| separated, source 150,000, target 50,000 | 110,000 | 8*M40

I'm using an old version of `t2t` (1.0.4). Does that matter?

**Comment by lukaszkaiser** [comment]
It's hard to tell. I'd suspect TF more than T2T, are you using TF 1.2.0? Is there nothing in the log at all, it just stops? Fails or hangs?

**Comment by cshanbo** [comment]
Yes, I'm using `tensorflow_gpu-1.2.0rc1-cp27-cp27mu-manylinux1_x86_64.whl`. 
There's nothing in the log at all, and the process just exits.

**Comment by cshanbo** [comment]
Hi @lukaszkaiser 
I'm using the `1.2.1` of tf right now, but the error remains. 
Actually, the model performs well, the only problem is the stopping phenomenon.
Is it any possibility that the model got `NaN` and break without any notification?

**Comment by justatest248** [comment]
Hi,

I've met the same issue, too.
I tried to track the error into `tensorflow`. Here is what I did:

Add logging info right  before and after [this line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L1010), where the training loop happens.

```
1007         while not mon_sess.should_stop():
1008           count += 1
1009           logging.info("tensorflow/contrib/learn/python/learn/estimators/estimator.py: count=%d"%(count))
1010           _, loss = mon_sess.run([model_fn_ops.train_op, model_fn_ops.loss])
1011           logging.info("tensorflow/contrib/learn/python/learn/estimators/estimator.py: loss=%f"%(loss))
```

After some steps of training, the logging info in line 1009 prints out normally, while line 1011 fails, like

```
 97 INFO:tensorflow:tensorflow/contrib/learn/python/learn/estimators/estimator.py: loss=2.381370
 98 INFO:tensorflow:tensorflow/contrib/learn/python/learn/estimators/estimator.py: count=650
 99 INFO:tensorflow:tensorflow/contrib/learn/python/learn/estimators/estimator.py: loss=2.073696
100 INFO:tensorflow:tensorflow/contrib/learn/python/learn/estimators/estimator.py: count=651
```

As a conclusion, this error happens in [line 1010](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L1010), where session.run happens.

I think it's better to let you know this error.

**Comment by lukaszkaiser** [comment]
We had problems with vocabulary sizes that lead to NaNs (thrown by the cross entropy on GPU when some labels were outside of range). This should be corrected now that we use the `Problem` class and the sizes are not twice in the code. So I'm closing this, but please reopen if you see it again!

**Comment by PrateekCoder** [comment]
@lukaszkaiser May I know how I can implement early stopping in transformers? I am trying to perform Machine Translation and I want to use early stopping but am not sure, how to do that.
